File: ruby/lib/cell.rb
# cell.rb
#
# vim: ft=ruby sw= ts=2
#
# cell class as basic element for a maze
#

class Cell

	attr_reader :row, :column
	attr_accessor :north, :south, :east, :west

	# constructor
	def initialize(row, column)
		@row, @column = row, column
		@links = {}
	end

	# add link (passage) to other cell
	def link(cell, bidi = true)
		@links[cell] = true
		cell.link(self, false) if bidi
	end

	# remove link (passage) to other cell
	def unlink(cell, bidi = true)
		@links.delete(cell)
		cell.unlink(self, false) if bidi
	end

	# returns all linked cells
	def links
		@links.keys
	end

	# checks whether a cell is linked
	def linked?(cell)
		@links.key? cell
	end

	# returns neighbors of the cell
	def neighbors
		list = []
		list << north if north
		list << south if south
		list << east if east
		list << west if west
		list
	end

end

File: js/lib/cell.js
// cell.js
// 
// vim: ft=javascript sw=4 ts=4
// 
// cell class as basic element for a maze
//

'use strict';

class Cell {

	constructor(row, column) {
		this.row = row;
		this.column = column;
		this.links = new Map();
		this.north = null;
		this.south = null;
		this.west = null;
		this.east = null;
	}

	// add link (passage) to other cell
	link(other, bidi = true) {
		this.links[other] = true;
		if (bidi) {
			other.link(this, false);
		}
	}

	// remove link (passage) to other cell
	unlink(other, bidi = true) {
		this.links.delete(other);
		if (bidi) {
			other.unlink(this, false);
		}
	}

	// returns all linked cells
	links() {
		return this.links.keys;
	}

	// checks whether a cell is linked
	linked?(other) {
		return this.links.has(other);
	}

	// returns neighbors of the cell
	neighbors() {
		var list = new Array(4);
		if (this.north) { list.push(this.north); }
		if (this.south) { list.push(this.south); }
		if (this.west) { list.push(this.west); }
		if (this.east) { list.push(this.east); }
	}

}

